Invalid Checkstyle format on successful linting

When using Checkstyle output format and the linting was successful, no output will be generated, thus the Checkstyle XML will be invalid

This will be an issue, when piping the output to another command (e.g. [`cs2pr`](https://github.com/staabm/annotate-pull-request-from-checkstyle))  
In my pipeline I run smth like this:
```
typoscript-lint --fail-on-warnings --format=xml --output=- | cs2pr
```

But this results in an error, because `cs2pr` requires a valid XML input:
```
typoscript-lint --fail-on-warnings --format=xml --output=- | cs2pr

Complete without warnings
Error: Expecting xml stream starting with a xml opening tag.

Error: Process completed with exit code 2.
```

--- a/composer.json
+++ b/composer.json
@@ -26,8 +26,9 @@
 		"symfony/filesystem": "^5.4 || ^6.4 || ^7.0",
 		"symfony/event-dispatcher": "^5.4 || ^6.4 || ^7.0",
 		"helmich/typo3-typoscript-parser": "^2.3",
-		"ext-json": "*"
-	},
+		"ext-json": "*",
+      	"ext-dom": "*"
+    },
 	"require-dev": {
 		"phpunit/phpunit": "^9.6.8",
 		"mikey179/vfsstream": "^1.6.11",

--- a/src/Linter/Report/Report.php
+++ b/src/Linter/Report/Report.php
@@ -52,6 +52,11 @@ class Report
         return $count;
     }
 
+    public function hasIssues(): bool
+    {
+        return $this->countIssues() > 0;
+    }
+
     /**
      * Returns the number of issues with a given severity for the entire report
      *

--- a/src/Linter/ReportPrinter/CheckstyleReportPrinter.php
+++ b/src/Linter/ReportPrinter/CheckstyleReportPrinter.php
@@ -3,7 +3,11 @@
 namespace Helmich\TypoScriptLint\Linter\ReportPrinter;
 
 use DOMDocument;
+use DOMElement;
+use DOMException;
 use Helmich\TypoScriptLint\Application;
+use Helmich\TypoScriptLint\Linter\Report\File;
+use Helmich\TypoScriptLint\Linter\Report\Issue;
 use Helmich\TypoScriptLint\Linter\Report\Report;
 use Symfony\Component\Console\Output\OutputInterface;
 
@@ -38,10 +42,23 @@ class CheckstyleReportPrinter implements Printer
      * Writes a report in checkstyle XML format.
      *
      * @param Report $report The report to print.
-     *
      * @return void
+     * @throws PrinterException
      */
     public function writeReport(Report $report): void
+    {
+        try {
+            $xml = $this->buildReportXMLDocument($report);
+            $this->output->write($xml->saveXML());
+        } catch (DOMException $error) {
+            throw new PrinterException('Could not generate checkstyle report: ' . $error->getMessage(),  $error);
+        }
+    }
+
+    /**
+     * @throws DOMException
+     */
+    private function buildReportXMLDocument(Report $report): DOMDocument
     {
         $xml = new DOMDocument('1.0', 'UTF-8');
 
@@ -49,30 +66,45 @@ class CheckstyleReportPrinter implements Printer
         $root->setAttribute('version', Application::APP_NAME . '-' . Application::APP_VERSION);
 
         foreach ($report->getFiles() as $file) {
-            $xmlFile = $xml->createElement('file');
-            $xmlFile->setAttribute('name', $file->getFilename());
-
-            foreach ($file->getIssues() as $issue) {
-                $xmlWarning = $xml->createElement('error');
-                $xmlWarning->setAttribute('line', $issue->getLine() !== null ? ((string)$issue->getLine()) : "");
-                $xmlWarning->setAttribute('severity', $issue->getSeverity());
-                $xmlWarning->setAttribute('message', $issue->getMessage());
-                $xmlWarning->setAttribute('source', $issue->getSource());
-
-                $column = $issue->getColumn();
-                if ($column !== null) {
-                    $xmlWarning->setAttribute('column', "" . $column);
-                }
-
-                $xmlFile->appendChild($xmlWarning);
-            }
-
-            $root->appendChild($xmlFile);
+            $root->appendChild($this->buildElementForFile($xml, $file));
         }
 
         $xml->appendChild($root);
         $xml->formatOutput = true;
 
-        $this->output->write($xml->saveXML());
+        return $xml;
+    }
+
+    /**
+     * @throws DOMException
+     */
+    public function buildElementForIssue(DOMDocument $xml, Issue $issue): DOMElement
+    {
+        $xmlWarning = $xml->createElement('error');
+        $xmlWarning->setAttribute('line', $issue->getLine() !== null ? ((string)$issue->getLine()) : "");
+        $xmlWarning->setAttribute('severity', $issue->getSeverity());
+        $xmlWarning->setAttribute('message', $issue->getMessage());
+        $xmlWarning->setAttribute('source', $issue->getSource());
+
+        $column = $issue->getColumn();
+        if ($column !== null) {
+            $xmlWarning->setAttribute('column', "" . $column);
+        }
+        return $xmlWarning;
+    }
+
+    /**
+     * @throws DOMException
+     */
+    public function buildElementForFile(DOMDocument $xml, File $file): DOMElement
+    {
+        $xmlFile = $xml->createElement('file');
+        $xmlFile->setAttribute('name', $file->getFilename());
+
+        foreach ($file->getIssues() as $issue) {
+            $xmlFile->appendChild($this->buildElementForIssue($xml, $issue));
+        }
+
+        return $xmlFile;
     }
 }

--- a/src/Linter/ReportPrinter/ConsoleReportPrinter.php
+++ b/src/Linter/ReportPrinter/ConsoleReportPrinter.php
@@ -40,6 +40,10 @@ class ConsoleReportPrinter implements Printer
      */
     public function writeReport(Report $report): void
     {
+        if (!$report->hasIssues()) {
+            return;
+        }
+
         $count = 0;
 
         $this->output->writeln('');

--- a/src/Linter/ReportPrinter/Printer.php
+++ b/src/Linter/ReportPrinter/Printer.php
@@ -13,5 +13,10 @@ use Helmich\TypoScriptLint\Linter\Report\Report;
 interface Printer
 {
 
+    /**
+     * @param Report $report
+     * @return void
+     * @throws PrinterException
+     */
     public function writeReport(Report $report): void;
 }

--- a/src/Linter/ReportPrinter/PrinterException.php
+++ b/src/Linter/ReportPrinter/PrinterException.php
@@ -1,0 +1,15 @@
+<?php
+declare(strict_types=1);
+
+namespace Helmich\TypoScriptLint\Linter\ReportPrinter;
+
+use Exception;
+use Throwable;
+
+class PrinterException extends Exception
+{
+    public function __construct(string $message, ?Throwable $previous = null)
+    {
+        parent::__construct($message, 0, $previous);
+    }
+}

--- a/src/Logging/CompactConsoleLogger.php
+++ b/src/Logging/CompactConsoleLogger.php
@@ -103,10 +103,10 @@ class CompactConsoleLogger implements LinterLoggerInterface
 
         if ($this->issueCount > 0) {
             $this->output->writeln("Completed with <comment>{$this->issueCount} issues</comment>");
-
-            $this->printer->writeReport($report);
         } else {
             $this->output->writeln("Complete <info>without warnings</info>");
         }
+
+        $this->printer->writeReport($report);
     }
 }

